### PR TITLE
graph/graphs/gen: add simple graph construction functions

### PR DESCRIPTION
**Pull request by kortschak**
~~WIP(notests)~~

Please take a look.

Updates #1582.

/cc @mattConn

### CODE REVIEW

**Comment by mattConn** [comment]
This looks good. `IDer` interface does clean things up.

**Comment by kortschak** [comment]
OK.  PTAL

I'm having second thoughts about checking for adjacent ID collision since this prevents multigraphs from being used to their fullest, but this can be discussed.

File: graph/graphs/gen/gen_example_test.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+	// strict graph star {
+	// 	// Node definitions.
+	// 	0;
+	// 	1;
+	// 	2;
+	// 	3;
+	// 	4;
+	// 	5;
+	// 	6;
+
```
Comment (mattConn) [inline graph/graphs/gen/gen_example_test.go:45]: I could not get these examples to execute until I commented out this newline and others which I'll mark below. After that, they all passed.

File: graph/graphs/gen/gen_example_test.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+	// strict digraph wheel {
+	// 	// Node definitions.
+	// 	0;
+	// 	1;
+	// 	2;
+	// 	3;
+	// 	4;
+	// 	5;
+	// 	6;
+
```
Comment (mattConn) [inline graph/graphs/gen/gen_example_test.go:78]: Comment out newline.

File: graph/graphs/gen/gen_example_test.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+	fmt.Printf("%s\n", b)
+
+	// Output:
+	// strict digraph path {
+	// 	// Node definitions.
+	// 	2;
+	// 	4;
+	// 	5;
+	// 	9;
+
```
Comment (mattConn) [inline graph/graphs/gen/gen_example_test.go:120]: Comment out newline.

File: graph/graphs/gen/gen_example_test.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+	fmt.Printf("%s\n", b)
+
+	// Output:
+	// strict digraph complete {
+	// 	// Node definitions.
+	// 	2;
+	// 	4;
+	// 	5;
+	// 	9;
+
```
Comment (mattConn) [inline graph/graphs/gen/gen_example_test.go:147]: Comment out newline.

File: graph/graphs/gen/gen_example_test.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
```
Comment (mattConn) [inline graph/graphs/gen/gen_example_test.go:187]: Comment out newline.

File: graph/graphs/gen/gen_example_test.go [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+	fmt.Printf("%s\n", b)
+
+	// Output:
+	// strict graph complete {
+	// 	// Node definitions.
+	// 	2;
+	// 	4;
+	// 	5;
+	// 	9;
+
```
Comment (mattConn) [inline graph/graphs/gen/gen_example_test.go:223]: Comment out newline.

File: graph/graphs/gen/gen_example_test.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+	// 	5 -> 6;
+	// 	6 -> 1;
+	// }
+}
+
+// IDSet is an IDer providing an explicit set of IDs.
+type IDSet []int64
+
+func (s IDSet) Len() int       { return len(s) }
+func (s IDSet) ID(i int) int64 { return s[i] }
```
Comment (mattConn) [inline graph/graphs/gen/gen_example_test.go:99]: I like this.

**Comment by kortschak** [comment]
And now with actual tests—that were written but not committed.

**Comment by mattConn** [comment]
LGTM

**Review by mattConn — DISMISSED** [review]

File: graph/graphs/gen/gen.go [hunk 8]
```diff
@@ -18,3 +22,122 @@ func abs(a int) int {
...
+	ID(int) int64
+}
+
+// Complete constructs a complete graph in dst using nodes with the given IDs.
+func Complete(dst NodeIDGraphBuilder, ids IDer) error {
+	for i := 0; i < ids.Len(); i++ {
+		uid := ids.ID(i)
+		u, new := dst.NodeWithID(uid)
+		if new {
+			dst.AddNode(u)
```
Comment (vladimir-ch) [inline graph/graphs/gen/gen.go:50]: Are you adding the node here to cover the case of `ids.Len() == 1`? Would it make sense to separate it as in the other functions?

File: graph/graphs/gen/gen_example_test.go [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+	"log"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/encoding/dot"
+	"gonum.org/v1/gonum/graph/graphs/gen"
+	"gonum.org/v1/gonum/graph/simple"
+)
+
+// IDRange is an IDer that provides a set of IDs in [First, Last].
+type IDRange struct{ First, Last int64 }
```
Comment (vladimir-ch) [inline graph/graphs/gen/gen_example_test.go:18]: This type and `IDSet` look convenient enough to be included in the `gen` package.

**Review by vladimir-ch — COMMENTED** [review]
With the two new interface this turned out much better than I expected!

File: graph/graphs/gen/gen_example_test.go (hunk 9 above)
Comment (kortschak) [reply graph/graphs/gen/gen_example_test.go:18]: Moved into package.

File: graph/graphs/gen/gen.go (hunk 8 above)
Comment (kortschak) [reply graph/graphs/gen/gen.go:50]: Done

**Comment by kortschak** [comment]
PTAL

**Review by vladimir-ch — APPROVED** [review]

**Pull request merged**